opm/parser/eclipse/EclipseState/Schedule/SegmentSet.cpp

```diff
@@ -0,0 +1,277 @@
+            copy->m_segments[i] = m_segments[i];
+        }
+        return copy;
+    }
+
+    void SegmentSet::segmentsFromWELSEGSKeyword(DeckKeywordConstPtr welsegsKeyword, DeckKeywordConstPtr nsegmxKeyword) {
+        // get the required information for multi-segment wells
+        const int nsegmx = nsegmxKeyword->getRecord(0)->getItem("NSEGMX")->getInt(0);
+        const int nlbrmx = nsegmxKeyword->getRecord(0)->getItem("NLBRMX")->getInt(0);
+
+        // for the first record, which provides the information for the top segment
+        // and information for the whole segment set
+        DeckRecordConstPtr record1 = welsegsKeyword->getRecord(0);
+        m_well_name = record1->getItem("WELL")->getTrimmedString(0);
+
+        // create a temporary vector to store all the segments information in a sparse way
+        // then compress the vector in a orderly way to m_segments
+        std::vector<SegmentPtr> new_segments;
+        new_segments.resize(nsegmx);
+
+        for (size_t i = 0; i < new_segments.size(); ++i) {
+            new_segments[i] = NULL;
+        }
+
+        const double meaningless_value = -1.e100; // meaningless value to indicate unspecified values
+
+        m_depth_top = record1->getItem("DEPTH")->getRawDouble(0);
+        m_length_top = record1->getItem("LENGTH")->getRawDouble(0);
+        m_length_depth_type = WellSegment::LengthDepthEnumFromString(record1->getItem("INFO_TYPE")->getTrimmedString(0));
+        m_volume_top = record1->getItem("WELLBORE_VOLUME")->getRawDouble(0);
+        m_comp_pressure_drop = WellSegment::CompPressureDropEnumFromString(record1->getItem("PRESSURE_COMPONENTS")->getTrimmedString(0));
+        m_multiphase_model = WellSegment::MultiPhaseModelEnumFromString(record1->getItem("FLOW_MODEL")->getTrimmedString(0));
+        m_x_top = record1->getItem("TOP_X")->getRawDouble(0);
+        m_y_top = record1->getItem("TOP_Y")->getRawDouble(0);
+
+        // the main branch is 1 instead of 0.
+        if (m_length_depth_type == WellSegment::INC) {
+            SegmentPtr top_segment(new Segment(1, 1, 0, 0., 0., meaningless_value, meaningless_value, meaningless_value,
+                                               m_volume_top, 0., 0., false));
+            new_segments[0] = top_segment;
+        } else if (m_length_depth_type == WellSegment::ABS) {
+            SegmentPtr top_segment(new Segment(1, 1, 0, m_length_top, m_depth_top, meaningless_value, meaningless_value,
+                                               meaningless_value, m_volume_top, m_x_top, m_y_top, true));
+            new_segments[0] = top_segment;
+        }
+
+        // read all the information out from the DECK first then process to get all the required information
+        for (size_t recordIndex = 1; recordIndex < welsegsKeyword->size(); ++recordIndex) {
+            DeckRecordConstPtr record = welsegsKeyword->getRecord(recordIndex);
+            int K1 = record->getItem("SEGMENT1")->getInt(0);
```

Comment:
Please do not use variable names `K1` and `K2` - it gives associations to the `z` coordinate.